### PR TITLE
[py] Add low-level sync API to use DevTools

### PR DESCRIPTION
**Pull request by p0deje**
### **User description**
Initial implementation of #13975. This provides a low-level API that:
1. establishes WebSocket connection to the driver;
2. use the connection to send CDP commands and receive results;
3. use the connection to subscribe to events;

Once this is merged, I'll work on the subsequent PRs that provide high-level API backed by CDP for now (network interception, log events, basic authentication, etc).

The implementation can also be used to send BiDi commands and subscribe to events once BiDi is implemented in Python (assuming we follow the same pattern as for CDP).

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Added `start_devtools` method in `webdriver.py` to establish a WebSocket connection and interact with DevTools.
- Implemented `WebSocketConnection` class to manage WebSocket connections, send and receive messages, and handle events.
- Added tests for DevTools WebSocket connection and console message logging.
- Added `websocket-client` dependency to `requirements.txt`.
- Enhanced `event_class` decorator to add `event_class` attribute to classes.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>generate.py</strong><dd><code>Add `event_class` attribute to event classes.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
py/generate.py

<li>Added <code>event_class</code> attribute to classes decorated with <code>event_class</code> <br>decorator.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13977/files#diff-197acd5304c91406cc4a12818c1694fe525f0efb718a12996516b082c29345f1">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>webdriver.py</strong><dd><code>Add `start_devtools` method for WebSocket connection to DevTools.</code></dd></summary>
<hr>
      
py/selenium/webdriver/remote/webdriver.py

<li>Imported <code>WebSocketConnection</code>.<br> <li> Added <code>start_devtools</code> method to establish WebSocket connection and <br>interact with DevTools.<br> <li> Initialized <code>_websocket_connection</code> attribute.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13977/files#diff-c87c150ed2a08f5293db58a06da8bf7f14c3a5582586c6270c1be661dfdd0985">+29/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>websocket_connection.py</strong><dd><code>Implement `WebSocketConnection` class for WebSocket communication.</code></dd></summary>
<hr>
      
py/selenium/webdriver/remote/websocket_connection.py

<li>Created <code>WebSocketConnection</code> class to manage WebSocket connections.<br> <li> Implemented methods for sending and receiving messages, handling <br>events, and managing connection state.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13977/files#diff-7db8ccff1907ae8f1040877b71b60dc4a77bd10ad6838e2eca856d2faf8d4794">+107/-0</a>&nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>devtools_tests.py</strong><dd><code>Add tests for DevTools WebSocket connection and console logging.</code></dd></summary>
<hr>
      
py/test/selenium/webdriver/common/devtools_tests.py

<li>Added tests for DevTools WebSocket connection and console message <br>logging.<br>

</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13977/files#diff-800b0102b563606add570f812f89944831fc639473425183366834ab7a754c0f">+38/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Dependencies
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>requirements.txt</strong><dd><code>Add `websocket-client` dependency.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
py/requirements.txt

- Added `websocket-client` dependency.

</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13977/files#diff-6aeed0b3fd7a115160945fdabc2be76056d10a8afdea93befd40d1fc05448e7c">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: py/selenium/webdriver/remote/webdriver.py [hunk 1]
```diff
@@ -1017,6 +1020,32 @@ def get_log(self, log_type):
         """
         return self.execute(Command.GET_LOG, {"type": log_type})["value"]
 
+    def start_devtools(self):
```
Comment (p0deje) [inline py/selenium/webdriver/remote/webdriver.py:1024]: I'm not sure about the naming here, any suggestions are welcome.

File: py/selenium/webdriver/remote/websocket_connection.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        self._start_ws()
+        self._wait_until(lambda: self._started)
+
+    def close(self):
+        self._ws_thread.join()
+        self._ws.close()
+        self._started = False
+        self._ws = None
+
+    def execute(self, command):
```
Comment (p0deje) [inline py/selenium/webdriver/remote/websocket_connection.py:52]: A general comment that applies to this whole PR - do we want to have typing properly implemented? I'm not familiar with the Python typing system so I did what was the easiest for me. I can re-iterate and add types if that's preferred.

File: py/selenium/webdriver/remote/webdriver.py (hunk 1 above)
Comment (diemol) [reply py/selenium/webdriver/remote/webdriver.py:1024]: Java calls this `getDevTools()`, I am not sure what is the idiomatic fit for Python.

File: py/selenium/webdriver/remote/websocket_connection.py (hunk 2 above)
Comment (diemol) [reply py/selenium/webdriver/remote/websocket_connection.py:52]: I would invest in typing for the BiDi implementation. Not much in the CDP one.

File: py/selenium/webdriver/remote/webdriver.py (hunk 1 above)
Comment (nir-tal-talkspace) [reply py/selenium/webdriver/remote/webdriver.py:1024]: to start a devtools session in Java ->  devTools.createSession();

Comment (p0deje) [reply py/selenium/webdriver/remote/webdriver.py:1024]: This method both _gets_ DevTools (loads the proper version) and _creates session_, then returns both the loaded devtools module and the established connection. `start_devtools` is the best possible name I came up with for these 2 operations. We could of course split them into 2 separate methods...

File: py/selenium/webdriver/remote/websocket_connection.py (hunk 2 above)
Comment (p0deje) [reply py/selenium/webdriver/remote/websocket_connection.py:52]: Types are already available in CDP, the question is should I add types to the `WebSocketConnection` class.

Comment (titusfortner) [reply py/selenium/webdriver/remote/websocket_connection.py:52]: I don't think this should be the priority right now. Focus on getting the hard parts implemented, and we can get help from others on the typing later.

File: py/selenium/webdriver/remote/webdriver.py (hunk 1 above)
Comment (titusfortner) [reply py/selenium/webdriver/remote/webdriver.py:1024]: for bidi this should be an implementation detail in the higher level APIs, so not too concerned about it.

**Review by titusfortner — APPROVED** [review]

**Comment by titusfortner** [comment]
@p0deje you can lint with `./go py:lint`

```
selenium/webdriver/remote/websocket_connection.py:47:38: F821 undefined name '_response_wait_timeout'
selenium/webdriver/remote/websocket_connection.py:78:19: F821 undefined name 'InternalError'
test/selenium/webdriver/common/devtools_tests.py:19:1: F401 'selenium.webdriver.common.by.By' imported but unused
test/selenium/webdriver/common/devtools_tests.py:20:1: F401 'selenium.webdriver.common.log.Log' imported but unused
test/selenium/webdriver/common/devtools_tests.py:21:1: F401 'selenium.webdriver.support.expected_conditions as EC' imported but unused
```

**Pull request merged**